llvm/lib/Frontend/HLSL/RootSignatureMetadata.cpp

```diff
@@ -538,6 +543,68 @@ Error MetadataParser::parseRootSignatureElement(mcdxbc::RootSignatureDesc &RSD,
   llvm_unreachable("Unhandled RootSignatureElementKind enum.");
 }
 
+Error validateDescriptorTableSamplerMixin(mcdxbc::DescriptorTable Table,
+                                          uint32_t Location) {
+  bool HasSampler = false;
+  bool HasOtherRangeType = false;
+  dxil::ResourceClass OtherRangeType;
+
+  for (const mcdxbc::DescriptorRange &Range : Table.Ranges) {
+    dxil::ResourceClass RangeType =
+        static_cast<dxil::ResourceClass>(Range.RangeType);
+
+    if (RangeType == dxil::ResourceClass::Sampler) {
+      HasSampler = true;
+    } else {
+      HasOtherRangeType = true;
+      OtherRangeType = RangeType;
+    }
+  }
+
+  // Samplers cannot be mixed with other resources in a descriptor table.
+  if (HasSampler && HasOtherRangeType)
+    return make_error<TableSamplerMixinError>(OtherRangeType, Location);
```

Comment:
Is it deliberate that we report the last non-sampler range type rather than the first? I think the logic would be simpler here if we just tracked the current type and errored if we ever see a sampler when we have a current type other than a sampler. Then we could just do an early return and we need fewer bookkeeping variables to keep track of. Something like (untested):
```c++
  dxil::ResourceClass CurrRC = dxil::ResourceClass::Sampler;
  for (const mcdxbc::DescriptorRange &Range : Table.Ranges) {
    if (Range.RangeType == dxil::ResourceClass::Sampler &&
        CurrRC != dxil::ResourceClass::Sampler)
      return make_error<TableSamplerMixinError>(CurrRC, Location);
    CurrRC = Range.RangeType;
  }

  return Error::success();
```